psl-core/src/main/java/org/linqs/psl/database/atom/OnlineAtomManager.java

```diff
@@ -129,6 +143,15 @@ public synchronized Boolean hasNewAtoms() {
         return (newRandomVariableAtoms.size() > 0) || (newObservedAtoms.size() > 0);
     }
 
+    /**
+     * Return the existing observed atoms.
+     */
+    public Set<GroundAtom> flushObservedAtoms() {
+        Set<GroundAtom> atoms = new HashSet<GroundAtom>(observedAtoms);
+        observedAtoms.clear();
+        return atoms;
```

Comment:
For all of these flushes, how about we just do a swaparoo (not a technical term) instead of a clear:
```
Set<GroundAtom> toReturn = observedAtoms;
observedAtoms = new HashSet<GroundAtom>();
return toReturn;
```